### PR TITLE
web: block-incremental parse layer for the streaming render, with its equivalence corpus and a benchmark; v0.9.57

### PR DESCRIPTION
**Pull request by cymoo**
Step 1 of #110 (landing order agreed there): the parse layer, its equivalence test, and a reproducible baseline. **The UI's rendering is unchanged** — `stream-blocks.js` is not imported by `chat.js` yet; that is step 3.

## What's in it

- **`lovia/web/static/js/stream-blocks.js`** — pure parse layer (no DOM). Decides which leading blocks of a streaming reply are final, renders each once, re-parses only the tail. Block-only lex with `space` tokens folded; final = ends with a blank line and two blocks follow; committed prefix re-verified by `startsWith` every flush; raw HTML blocks and link-reference definitions (as marked itself classifies them, via `lexer.tokens.links`) fall back to full render for the rest of the reply. Heading ids stay on: a persistent `marked.Parser` de-duplicates across committed blocks and the tail renders through a copy of its slugger state. Committed blocks are never un-committed when the stable count drops; the invariant (I2) is in the module header.
- **`tests/web/js/stream_blocks.test.mjs`** — every prefix of 18 edge-case documents and of README / README-zh / AGENTS / docs/architecture, asserting `join(committed) + tail === marked.parse(prefix)`, then 10 deterministic random flush histories per document; plus mismatch-reset and permanent-fallback contracts. Documents without HTML / link defs must run 100 % incremental (a green run can't hide behind fallbacks). node's built-in runner, wired into the existing `typecheck-web` job — no npm packages. 27 tests, ~14 s (`EVERY_PREFIX=1` for the exhaustive run). Runs with `mangle: false` because marked v4 randomizes mailto entity encoding per parse (found by CI through AGENTS.md; pinned as `emailInOpenCode`).
- **`scripts/bench_stream_render.py`** — headless-Chromium benchmark against a running `lovia web`. `render` mode: per-flush cost of the exact `flushRender` pipeline (markdown → sanitize → DOM → hljs → copy buttons/labels → mermaid scan) by size and shape, p50/p95/max. `live` mode: renders, first-paint latency, rAF long frames for one streamed reply until the run settles (`aria-busy`), stub model. Output is markdown.
- `chat.js`: `highlightCode` exported for the benchmark. `types.d.ts`: `MarkedStatic` gains `Lexer` / `Parser` / `lexer` / `defaults`. AGENTS.md: command-table row for the JS checks.

## Baseline (current renderer, this machine)

Full flush, p50 ms, hljs cache warm for the mixed document:

| document | size | blocks | parse | purify | DOM+layout | hljs | chrome | flush p50 / p95 |
| --- | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| mixed | 4 KB | 105 | 0.4 | 0.9 | 1.6 | 0.4 | 0.5 | **3.5** / 7.0 |
| mixed | 12 KB | 315 | 1.2 | 2.6 | 4.5 | 1.4 | 1.6 | **10.7** / 13.7 |
| mixed | 24 KB | 620 | 2.4 | 5.1 | 8.8 | 2.5 | 3.1 | **20.5** / 27.8 |
| mixed | 48 KB | 1220 | 4.4 | 10.2 | 17.9 | 5.1 | 6.4 | **42.8** / 46.2 |
| one growing fenced code block | 48 KB | 1 | 0.3 | 0.4 | 5.7 | 30.5 | 2.9 | **31.4** / 36.4 |
| one growing table | 48 KB | 1 | 5.4 | 12.7 | 22.0 | 0.1 | 0.2 | **39.9** / 44.1 |
| one paragraph, no blank lines | 48 KB | 1 | 0.7 | 0.2 | 20.6 | 0.0 | 0.0 | **21.3** / 22.0 |

Live, 48 KB at ~10 ms per delta: 606 renders over 49 s (~13/s × 43 ms ≈ 55 % of the main thread), 2 long frames, max gap 50 ms, first text at 134 ms.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: lovia/web/static/js/stream-blocks.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+//       non-blank line, which was already present.
+//
+// The stable count can therefore *drop* below what is committed (a setext
+// underline, or the lone "-" of a list item still being typed, merges the two
+// blocks after a committed one into one). That is bookkeeping, not
+// instability: committed blocks are never un-committed. Un-committing would
+// also have to roll back cross-block parser state (heading-id de-duplication),
+// which is exactly what went wrong when it was tried.
+
+const LINK_DEF = /^ {0,3}\[[^\]]+\]:/m;
```
[the remark this answers is not in the document]
Comment (cymoo) [reply lovia/web/static/js/stream-blocks.js]: Fixed in f73f987: `lexBlocks` now reports whether marked itself recorded a definition (`lexer.tokens.links`) and the regex is gone, so `[x]: …` inside a fence or on the line after a paragraph (absorbed as text) no longer triggers the fallback. Both cases are pinned in the corpus (`defInCode`, `defAfterParagraph`) and must stay 100 % incremental.

File: scripts/bench_stream_render.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    let t0 = performance.now(); const html = marked.parse(src); t.parse.push(performance.now() - t0);
+    t0 = performance.now(); const clean = DOMPurify.sanitize(html); t.purify.push(performance.now() - t0);
+    t0 = performance.now();
+    const tmpl = document.createElement('template'); tmpl.innerHTML = clean;
+    body.replaceChildren(tmpl.content); body.offsetHeight;
+    t.dom.push(performance.now() - t0);
+    t0 = performance.now(); util.highlightIn(body); body.offsetHeight; t.hljs.push(performance.now() - t0);
+    // The whole flush, on text the staged hljs pass above has not cached.
+    const src2 = grow ? src + '.' : src;
+    t0 = performance.now(); util.renderMarkdownInto(body, src2); util.highlightIn(body); body.offsetHeight;
```
[the remark this answers is not in the document]
Comment (cymoo) [reply scripts/bench_stream_render.py]: Fixed in f73f987: the probe now imports `highlightCode` (exported from chat.js for this) and `renderMermaid` and times exactly what `flushRender` runs, with a separate `chrome` stage (copy buttons + labels + mermaid scan). It moves the 48 KB mixed flush from 34.9 to 42.8 ms p50 — a real ~15 % the old baseline was missing. PR description updated.

File: scripts/bench_stream_render.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+def bench_live(page, window_s: float) -> None:
+    page.evaluate(LIVE_PROBE)
+    page.fill("#prompt", "go")
+    page.keyboard.press("Enter")
+    t0 = time.time()
+    done = None
+    while time.time() - t0 < window_s:
+        if done is None and page.query_selector("#followups .followup"):
+            done = time.time() - t0
+        time.sleep(0.1)
```
[the remark this answers is not in the document]
Comment (cymoo) [reply scripts/bench_stream_render.py]: Fixed in f73f987: live mode now watches `#transcript[aria-busy]`, which chat.js sets to `true` on stream start and back to `false` in `exitStreamingUI` when the run settles — independent of follow-up generation.
